Scripts/Runtime/Data/VirtualData/VirtualData.cs

```diff
@@ -143,35 +158,156 @@ internal sealed override JobHandle ConsolidateForFrame(JobHandle dependsOn)
         //*************************************************************************************************************
 
         [BurstCompile]
-        private struct ConsolidateLookupJob : IJob
+        private struct ConsolidateLookupJob : IAnvilJob
         {
             private UnsafeTypedStream<TInstance> m_Pending;
-            private DeferredNativeArray<TInstance> m_Iteration;
-            private UnsafeParallelHashMap<TKey, TInstance> m_Lookup;
+            private DeferredNativeArray<TInstance> m_IterationTarget;
+            private UnsafeParallelHashMap<VDContextID, TInstance> m_Lookup;
+            private VDLookupReader<bool> m_CancelledLookup;
+            //TODO: DISCUSS - Split this out into separate data 
+            private DeferredNativeArray<TInstance> m_CancelledIterationTarget;
 
             public ConsolidateLookupJob(UnsafeTypedStream<TInstance> pending,
-                                        DeferredNativeArray<TInstance> iteration,
-                                        UnsafeParallelHashMap<TKey, TInstance> lookup)
+                                        DeferredNativeArray<TInstance> iterationTarget,
+                                        UnsafeParallelHashMap<VDContextID, TInstance> lookup,
+                                        VDLookupReader<bool> cancelledLookup,
+                                        DeferredNativeArray<TInstance> cancelledIterationTarget)
             {
                 m_Pending = pending;
-                m_Iteration = iteration;
+                m_IterationTarget = iterationTarget;
                 m_Lookup = lookup;
+                m_CancelledLookup = cancelledLookup;
+                m_CancelledIterationTarget = cancelledIterationTarget;
+            }
+
+            public void InitForThread(int nativeThreadIndex)
+            {
             }
 
             public void Execute()
             {
+                //Clear previously consolidated 
                 m_Lookup.Clear();
-                m_Iteration.Clear();
+                m_IterationTarget.Clear();
+                m_CancelledIterationTarget.Clear();
 
-                NativeArray<TInstance> iterationArray = m_Iteration.DeferredCreate(m_Pending.Count());
-                m_Pending.CopyTo(ref iterationArray);
+                //Get the new counts
+                int pendingCancelledCount = m_CancelledLookup.Count();
+                int pendingCount = m_Pending.Count();
```

Comment:
> Why do this? If you have to iterate anyway why not just interate and do the work?

We don't have to iterate, the add/removes are what are iterated. We just need to copy all the contents of the lookup into a flat array after for load balancing and this provides a fast way to do that.

> Can you use removeSwapBack to cull all the cancelled tasks right off the update list and then perform your pending adds to minimize growth/shrink required?

Yes, we'll always process removes first and then adds. The collection will only grow, but never shrink capacity.

> Can you record the index of each task to be cancelled and pass that into the iterator for the Job. On each iteration it checks to see if a given index should be skipped and skips that job for update?

Possibly, I'll explore that.

Reply:
This was reworked to be more efficient. We explored a bunch of different options but ran into safety/usability issues on the fastest approach and the others were too slow. 

We've gone from about 10 to 13ms for 100,000 elements down to 3ms so that's pretty decent.